design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,471 @@
+- Only a single repository can exist at a particular path in the backend storage location. Users opting to use separate repositories for separate applications are recommended to use unique path prefixes for each repository. For example, the repository for `mysql-deploy` deployment in `mysql` namespace on S3 storage bucket called `test-bucket` could be created at the location `s3://test-bucket/<UUID of mysql namespace>/mysql-deploy/`.
+- Accessing the repository requires location and credential information similar to a Kanister Profile CR and a unique password used by Kopia during [encryption](https://kopia.io/docs/features/#end-to-end-zero-knowledge-encryption) along with a unique path prefix as mentioned above.
+- In the first iteration, users will be required to provide the location and repository information to the controller during the creation of the repository server in the form of Kubernetes Secrets. Future iterations will allow users to use a Key Management Service of choice.
+
+### Kopia Repository Server
+
+- A Kopia Repository Server allows Kopia clients proxy access to the backend storage location through it.
+- At any time, a repository server can only connect to a single repository. Due to this a separate instance of the server will be used for each repository.
+- In Kanister, the server will comprise a K8s `Pod`, `Service` and a `NetworkPolicy`.
+- The pod will execute the Kopia server process exposed to the application via the K8s service and the network policy.
+- Accessing the server requires the service name, a server username, and a password without any knowledge of the backend storage location.
+- To authorize access, a list of server usernames and passwords must be added prior to starting the server.
+- The server also uses TLS certificates to secure incoming connections to it.
+- Kanister users can configure a repository via a newly added Custom Resource called `RepositoryServer` as described in the following section.
+
+#### Custom Resource Definition
+
+This design proposes a new Custom Resource Definition (CRD) named `RepositoryServer`, to represent a Kopia repository server.
+It is a namespace-scoped resource, owned and managed by a new `RepositoryServerController` controller in the `kanister` namespace.
+As mentioned above, the CRD offers a set of parameters to configure a Kopia repository server.
+
+To limit the new controller's RBAC scope, all the `RepositoryServer` resources are created in the `kanister` namespace.
+
+A sample `RepositoryServer` resource created to interact with the Kopia repository of workloads running in the `monitoring` namespace looks like this:
+
+```yaml
+apiVersion: cr.kanister.io/v1alpha1
+kind: RepositoryServer
+metadata:
+   name: repository-monitoring
+   namespace: kanister
+   labels:
+      repo.kanister.io/target-namespace: monitoring
+spec:
+   location:
```

Comment:
As mentioned in [this comment](https://github.com/kanisterio/kanister/pull/1645#discussion_r979895726), can we rename the `location` field here?

The Location struct for Profile CR [in pkg/apis/cr/v1alpha1/types.go](https://github.com/kanisterio/kanister/blob/dab83e87b16e7eeb73645ed0768450fd978ee2b7/pkg/apis/cr/v1alpha1/types.go#L255) is conflicting with this location field here

Reply:
@shlokchaudhari9 Yes, I think `storage` might work here as well. I'll change it 👍🏼